internal/engine/experiment/echcheck/handshake.go

```diff
@@ -0,0 +1,60 @@
+package echcheck
+
+import (
+	"context"
+	"crypto/rand"
+	"crypto/tls"
+	"github.com/apex/log"
+	"github.com/ooni/probe-cli/v3/internal/measurexlite"
+	"github.com/ooni/probe-cli/v3/internal/model"
+	"github.com/ooni/probe-cli/v3/internal/netxlite"
+	utls "gitlab.com/yawning/utls.git"
+	"net"
+	"time"
+)
+
+const echExtensionType uint16 = 0xfe0d
+
+func handshake(ctx context.Context, conn net.Conn, zeroTime time.Time, address string, sni string) *model.ArchivalTLSOrQUICHandshakeResult {
+	return handshakeWithExtension(ctx, conn, zeroTime, address, sni, []utls.TLSExtension{})
+}
+
+func handshakeWithEch(ctx context.Context, conn net.Conn, zeroTime time.Time, address string, sni string) *model.ArchivalTLSOrQUICHandshakeResult {
+	payload, err := generateGreaseExtension(rand.Reader)
+	if err != nil {
+		panic("failed to generate grease ECH: " + err.Error())
+	}
+
+	var utlsEchExtension utls.GenericExtension
+
+	utlsEchExtension.Id = echExtensionType
+	utlsEchExtension.Data = payload
+
+	return handshakeWithExtension(ctx, conn, zeroTime, address, sni, []utls.TLSExtension{&utlsEchExtension})
+}
+
+func handshakeWithExtension(ctx context.Context, conn net.Conn, zeroTime time.Time, address string, sni string, extensions []utls.TLSExtension) *model.ArchivalTLSOrQUICHandshakeResult {
+	tlsConfig := genTLSConfig(sni)
+
+	handshakerConstructor := newHandshakerWithExtensions(extensions)
+
+	start := time.Now()
+
+	tracedHandshaker := handshakerConstructor(log.Log, &utls.HelloFirefox_65)
```

Comment:
It seemed like the most appropriate choice from the list of [options](https://github.com/refraction-networking/utls/blob/v1.2.0/u_common.go#L158). `HelloRandomized` would not work because we need to use the same format for the control and target tests, `HelloGolang` does not allow for custom extensions, and `HelloCustom` requires manual construction of the Client Hello message. 

I've changed this to `HelloFirefox_Auto` in [405ba12](https://github.com/ooni/probe-cli/pull/970/commits/405ba12d43e1af197f3f9708744b80bd915815be) to make it more generic.

Reply:
Thank you for the clarification! :pray:  I wasn't aware that HelloGolang does not allow for custom extensions.